### ISSUE TITLE
Update from KlipperScreen doesn't restart Spoolman service

### ISSUE THREAD

**Issue by rdany62** [Donkie/Spoolman#247]
Thanks Donkie for Spoolman and that feature 😊

**Describe the bug**
When hitting update Spoolman in KlipperScreen the service doesn't restart.
And the update button remains visible éventuellement if no new Spoolman available (0.15 to 0.15 by exemple)

**To Reproduce**
Click the update button, Spoolman service is stopped, Spoolman is downloaded, Spoolman venv folder is deleted, but Spoolman doesn't seem to be reinstalled (venv not created, service unit still exists in /etc/système/system), service can't start.

**Expected behavior**
Spoolman updated and service restarted After the update 

**Screenshots**

![IMG_20240102_124340.jpg](https://github.com/Donkie/Spoolman/assets/66699240/c6f8a27b-dc34-43b6-bdf3-028794627097)
(Not the best quality for that one sorry)

![IMG_20240102_124431.jpg](https://github.com/Donkie/Spoolman/assets/66699240/db861827-2574-45ac-abc2-c3e259044dbb)

![IMG_20240102_130840.jpg](https://github.com/Donkie/Spoolman/assets/66699240/7d0a5ba8-f3f4-4573-a790-2a2aa469cc37)

**Desktop (please complete the following information):**
 - OS: Debian 12, klipper 0.12.0-31, moonraker 0.8.0-243, Spoolman 0.15

**Additional context**
Permission problem on ~/Spoolman/scripts files, chmod +x on install_debian.sh and start.sh, run the install script, the Spoolman-venv is created and the service starts without problem.

![IMG_20240102_130900.jpg](https://github.com/Donkie/Spoolman/assets/66699240/698f614b-66f0-44b1-ae2c-45eabc6812d6)

![IMG_20240102_131035.jpg](https://github.com/Donkie/Spoolman/assets/66699240/6d5ae882-aebd-40ab-a5f2-d7891c6e076b)

![IMG_20240102_132142.jpg](https://github.com/Donkie/Spoolman/assets/66699240/84b2a5f0-bc37-4670-8806-2d44d342edc4)

Hope that helps 😉

A bug related to KlipperScreen or moonraker I think: when the Spoolman service is stopped KlipperScreen crashes.

**Comment by rdany62** [comment]
I see I uploaded a wrong photo (the 1st one), sorry. Not related and can't modify with the app on my phone 😁

**Comment by Donkie** [comment]
I updated the spoolman.zip yesterday in the latest release to try fix this issue. Can you check if it works now?

**Comment by rdany62** [comment]
![IMG_20240103_125524.jpg](https://github.com/Donkie/Spoolman/assets/66699240/235de28e-59a6-463a-a5cb-8393f9af5a2d)

Juste tried but still have permission error, sorry

Or it didn't downloaded the zip and moonraker decompressed the old one (is it cached ?)

**Comment by rdany62** [comment]
Downloaded successfully with wget

![IMG_20240103_130257.jpg](https://github.com/Donkie/Spoolman/assets/66699240/ac1e63b8-41f0-4dfd-b407-64055f955e81)

**Comment by rdany62** [comment]
.venv doesn't exists anymore, I reinstall Spoolman

Updating via KlipperScreen/Moonraker doesn't set the permissions correctly, unzipping manually yes.

![IMG_20240103_131758.jpg](https://github.com/Donkie/Spoolman/assets/66699240/e688fbd2-e0e0-4264-8cbe-4af4f20aecf8)

**Comment by Arksine** [comment]
So, a couple of things on this:

- The `web_client` type does not support service management.  Its intended for front ends like Mainsail and Fluidd
- Unfortunately the permission issue is a [bug in Python](https://github.com/python/cpython/issues/59999).  The good news is that there is a workaround for it, which I intend to push to Moonraker.

Generally speaking Moonraker's `update_manager` only supports services updated through git.  For the time being, I think the best I can provide is the ability to update without changing permissions.  Restarting the service will need to be done manually.  If you add `spoolman` to `~/printer_data/moonraker.asvc` the Moonraker you will be able to use the frontend to perform a restart.

**Comment by rdany62** [comment]
Hello,

The zip file permissions issue is corrected but Spoolman still doesn't get updated, it seems it is only downloaded and extracted.
![IMG_20240120_085100.jpg](https://github.com/Donkie/Spoolman/assets/66699240/206cab3c-c7f0-49d4-81d5-e3cf94097ed6)


I have to launch the install_debian scripts from terminal, the venv folder is erased when the zip file is extracted.
![IMG_20240120_085405.jpg](https://github.com/Donkie/Spoolman/assets/66699240/12f384e2-62fa-4911-a6ba-a4a512531d79)


I added Spoolman to moonraker.asvc but still can't restart the service from KlipperScreen. Maybe because of the 'web' type in moonraker.conf:
![IMG_20240120_085748.jpg](https://github.com/Donkie/Spoolman/assets/66699240/ab3977dc-9c15-4011-b423-00ad4187c9e1)

I noticed something else, when Spoolman service is stopped and I try to go to the Spoolman menu KlipperScreen crashes:
![IMG_20240120_085259.jpg](https://github.com/Donkie/Spoolman/assets/66699240/61443bb9-3f3d-4279-b37c-c9f46b0ff5bb)

**Comment by Arksine** [comment]
> The zip file permissions issue is corrected but Spoolman still doesn't get updated, it seems it is only downloaded and extracted.

Correct.  Downloading and extracting is all the update manager's `web` type is capable of.  That said, I have been working on bringing up the undocumented `zip` type which can download a zipped distribution, install python and/or system package dependencies (this doesn't apply to Spoolman given the way the install is done), and restart the systemd service.  I plan to merge next week.  FWIW, you can prevent the update_manager from nuking your `.venv` folder by adding `.venv` to the `perisistent_files` option.  When I have merged the relevant changes I will include a full config section that should work.

> The 'update' button is always active, maybe because 'v.0.15.0' is compared to '0.15.0'. At the moment it helps since I can try the update process.

This is indeed an error in `release_info.json`.  The version reported needs to use the same version scheme as the release on github.  The release_info file reports the tag `v0.15.0`, whereas the release on GitHub is `0.15.0`.  FWIW, when testing I manually modify `release_info.json`, then restart Moonraker to "fake" a lower version.

**Comment by Arksine** [comment]
Ok, I have merged the changes.  The following configuration should work:
```
[update_manager Spoolman]
type: zip
channel: stable
repo: Donkie/Spoolman
path: ~/spoolman
persistent_files:
  .venv
  .env
managed_services: Spoolman
```

The above assumes that Spoolman has been installed using the `install-debian.sh` script.  The systemd service unit is case sensitive, so the section header and value for the `managed_service` option must match (ie: `Spoolman` instead of `spoolman`).  In addition `Spoolman` must be added to `~/printer_data/moonraker.asvc` to grant Moonraker permission to restart the service.  This too is case sensitive.

The above configuration will prevent the update manager from nuking the virtualenv and spoolman's environment file if it exists.  It can't resolve the issue with the mismatched version schemes, that is something Donkie will need to fix in the next release.

**Comment by rdany62** [comment]
@Arksine 

I tried the config section, Spoolman was already added to moonraker.asvc

I can now restart the Spoolman service and download and extract the spoolman.zip file from KlipperScreen.

But thé install_debian.sh script isn't launched, I think this is needed to update the server modules.

I tried to add install_script: scripts/install_debian.sh but it's still not launched.

And this script isn't fully unattented this may lock KlipperScreen because there is way to answer y or n

Please note that I installed Spoolman to ~/Spoolman directory, not to ~/spoolman. 

![IMG_20240121_184752.jpg](https://github.com/Donkie/Spoolman/assets/66699240/f8cb05ff-b1e5-4d92-afd1-17edc606e4a7)

@Donkie 
I tried to create a git section for Spoolman, this is really not as easy. I installed npm using nvm, npm install and npm run build but I get error without code and nothing in the log. I will continue to try, as it may be interesting to have it set up and working.

**Comment by Arksine** [comment]
> But thé install_debian.sh script isn't launched, I think this is needed to update the server modules.

Moonraker does not run scripts during update.   The `intall_script` option is deprecated, as its only purpose to provide system packages which Moonraker parsed out.  From the documentation:

```ini
install_script:
#  *** DEPRICATED FOR NEW CONFIGURATIONS - USE the 'system_dependencies' OPTION ***
#
#  The file location, relative to the repository, for the installation script
#  associated with this application.  Moonraker will not run this script, instead
#  it will parse the script searching for new "system" package dependencies that
#  require installation.  Packages in the script must be defined as follows for
#  Moonraker to successfully parse them:
#      PKGLIST="packagename1 packagename2 packagename3"
#      PKGLIST="${PKGLIST} packagename4 packagename5"
#
#  Note that the "packagenameX" items in the example above should be the names
#  of valid system packages.  The second line in the example is optional and
#  additional lines in the same format may be added.
#
#  The default is no install script.
```

If an application needs Python dependencies installed they can be provided through a requirements file.  Likewise, if it needs to install system packages there is an option to specify a `system_dependencies` file.  If an application needs to run a script to complete the update then it needs to perform that action itself, Moonraker will not do that.

**Comment by rdany62** [comment]
@Arksine thanks for your clarifications, I will check my config next week I have other things that launch a script when updating 😉

**Comment by Donkie** [comment]
Thanks for the work @Arksine . So to be clear, I just need to add "v" as a prefix to the release titles from now on and it should work correctly with the zip type?

**Comment by Arksine** [comment]
> Thanks for the work @Arksine . So to be clear, I just need to add "v" as a prefix to the release titles from now on and it should work correctly with the zip type? Would it help if I add v to the current release also?

Yes, that would work and match the scheme used by the front-ends.  It probably wouldn't hurt to add "v" to the current release, the next time Moonraker refreshes remote state it should pull in the new version.

> Also I'm still not quite sure how I should handle installing python dependencies..

If you need Moonraker to update Python dependencies it needs a python requirements file.  We would then change the configuration to look something like this:

```
[update_manager Spoolman]
type: zip
channel: stable
repo: Donkie/Spoolman
path: ~/spoolman
virtualenv: .venv
requirements: relpath/to/requirements.txt
persistent_files:
  .venv
  .env
managed_services: Spoolman
```

The `virtualenv` option can be relative to the `path` option or absolute since its common for venvs to exist outside of the application folder.  The `requirements` option must be in the `path`.  Note that the above configuration would result in an error right now because Spoolman doesn't deliver a requirements file, although it would probably be fine to add an empty one.

> Edit: it may be harder to update Spoolman if moonraker doesn't launch update script anymore. Maybe we can add a check which will launch such script when the service gets restarted just right after the update ?

To be clear, Moonraker's update manager has never ran a script during an update.  Running scripts introduces unpredictable behavior, therefore Moonraker avoid it and hands that off to the application itself.

**Comment by Donkie** [comment]
Ok got it, thanks for the details. I could probably remake the publishing flow so that the zip contains a compiled requirements.txt by PDM. Would mean people don't need PDM installed also which is a bonus.

**Comment by Arksine** [comment]
> Ok got it, thanks for the details. I could probably remake the publishing flow so that the zip contains a compiled requirements.txt by PDM. Would mean people don't need PDM installed also which is a bonus.

YW.  FWIW, I noticed during the install that a bunch of Python dependencies were installed via system packages.  These would automatically get updated by Moonraker when the user chooses to update the system, but typically prebuilt packages are pinned so it shouldn't affect anything.

**Comment by Arksine** [comment]
One other thing...when I get a chance I intend to look unto adding support for updating docker containers.  I haven't researched it yet but I don't imagine it would be too difficult, and something like that might be a good alternative Spoolman.

**Comment by rdany62** [comment]
@Arksine Ok, thanks you very much 😉


I tried to export package liste using pdm:
![IMG_20240121_202802.jpg](https://github.com/Donkie/Spoolman/assets/66699240/76ec395a-08ca-4cca-ac1e-2726f74aeefa)

I am not comfortable with the warning.

It generates the following file:
![IMG_20240121_202824.jpg](https://github.com/Donkie/Spoolman/assets/66699240/6674db76-13ee-4890-9c2f-ebe43326a52)


I can't test if moonraker updates Spoolman and the dependancies because the requirements.txt file will be erased when the zip file will be downloaded.

I prepared 2 requirements.txt files.
Are they okay or only the second one will work ?

The 1st:
![IMG_20240121_203412.jpg](https://github.com/Donkie/Spoolman/assets/66699240/105f27d4-85ef-4fd0-be25-41c26f4a7017)

The 2nd:
![IMG_20240121_203524.jpg](https://github.com/Donkie/Spoolman/assets/66699240/a36717a5-fa0c-484f-b929-2f1b4391dd61)

Hope this helps.

**Comment by Arksine** [comment]
> I prepared 2 requirements.txt files.
Are they okay or only the second one will work ?

The `requirements.txt` generated by pdm is what you need.  It looks like you tried to reformat it into the package list Moonraker parses from the install script...those are system packages (ie: packages installed via `apt-get`).  The requirements file contains python packages.

> I can't test if moonraker updates Spoolman and the dependancies because the requirements.txt file will be erased when the zip file will be downloaded.

Correct, you won't be able to test it until the package hosted on GitHub includes the requirements file.  Moonraker's procedure for updating Python requirements is as follows:
- Read out the current requirements file
- Download the package and overwrite existing files
- Read out the new requirements file
- Compare the two, run `.venv\bin\pip install -r <requreiments>` if a package version has changed or new packages are added

**Comment by rdany62** [comment]
Thanks you @arksine

If Spoolman can be updated without building the client each time nor running pdm it will be great, installation using pdm takes something like 12 or 15 minutes on my system. Simply download and uncompress a zip, and update python packagés ... Nice !

@Donkie 
Extracting packagé list from pdm.lock seems to be simple 😉

One more question, does moonraker support requirements.txt files with hashes ?

**Comment by Arksine** [comment]
> One more question, does moonraker support requirements.txt files with hashes ?

Yes, the requirements file may contain comments prefixed with `#`.

**Comment by rdany62** [comment]
@Arksine 

Sorry, I was speaking about hashes of the modules
We can obtain a requirements.txt with modules hashes with:
pdm export -o requirements.txt

**Comment by rdany62** [comment]
Hi,

The local zip version (v0.15.0) equals now the remote zip version.

We will see on n'est zip version if the update process works 😉

**Comment by rdany62** [comment]
Hello, sorry for time since last message.

I updated from v0.15.0 to v0.16.0, files were not readable nor writeable. chmodded all files corrected the problem 

Tried today to update from v0.16 0 to v0.16.1, the scripts are not executable.

![Screenshot_2024-03-05-17-47-51-293_com.sonelli.juicessh.jpg](https://github.com/Donkie/Spoolman/assets/66699240/be6c4cba-14e2-4a9c-86db-847dde155539)

chmod +x the files in scripts folder and ran thé install script, because I deleted thé persistent_files in moonraker.conf:
persistent_files:
	.venv
	.env

Now if Spoolman service is stopped KlipperScreen doesn't crash.
